File: macos/Tests/ElevateCoreTests/Support/Jwt.swift
import Foundation

/// Builds a JWT-shaped token carrying the given claims, so a test can drive claim reads.
enum Jwt {
    static func with(_ claims: [String: Any]) -> String {
        let body = try! JSONSerialization.data(withJSONObject: claims)
        let b64 = body.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .trimmingCharacters(in: CharacterSet(charactersIn: "="))
        return "eyJhbGciOiJub25lIn0.\(b64).sig"
    }

    /// A token whose `wids` names the given directory role template ids.
    static func withRoles(_ roleTemplateIds: String...) -> String { with(["wids": roleTemplateIds]) }

    /// A token whose `groups` names the given group object ids.
    static func withGroups(_ groupIds: String...) -> String { with(["groups": groupIds]) }
}

File: windows/tests/Elevate.Core.Tests/Support/Jwt.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Elevate.Core.Tests.Support;

/// <summary>Builds a JWT-shaped token carrying the given claims, so a test can drive claim reads. Port of the Swift <c>Jwt</c>.</summary>
public static class Jwt
{
    public static string With(Action<JsonObject> claims)
    {
        ArgumentNullException.ThrowIfNull(claims);
        var payload = new JsonObject();
        claims(payload);
        return $"eyJhbGciOiJub25lIn0.{Base64Url(JsonSerializer.SerializeToUtf8Bytes(payload))}.sig";
    }

    /// <summary>A token whose <c>wids</c> names the given directory role template ids.</summary>
    public static string WithRoles(params string[] roleTemplateIds) =>
        With(c => c["wids"] = new JsonArray([.. (roleTemplateIds ?? []).Select(id => (JsonNode)id!)]));

    /// <summary>A token whose <c>groups</c> names the given group object ids.</summary>
    public static string WithGroups(params string[] groupIds) =>
        With(c => c["groups"] = new JsonArray([.. (groupIds ?? []).Select(id => (JsonNode)id!)]));

    private static string Base64Url(byte[] bytes) =>
        Convert.ToBase64String(bytes).Replace('+', '-').Replace('/', '_').TrimEnd('=');
}
